include/revng/Pipes/Lift.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <array>
+
+#include "llvm/IR/LegacyPassManager.h"
+
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/InputOutputContract.h"
+#include "revng/Pipeline/InspectableLLVMContainer.h"
+#include "revng/Pipeline/LLVMContainer.h"
+#include "revng/Pipeline/LLVMPipe.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Pipes/RevngPipes.h"
+#include "revng/Pipes/TemporaryFile.h"
+
+extern llvm::cl::opt<unsigned long long> BaseAddress;
+
+namespace Pipeline {
+
+class LiftPipe {
+public:
+  static constexpr auto Name = "Lift Pipe";
+  std::array<AtomicContract, 1> getContract() const {
+    return { AtomicContract(Binary, Exactness::Exact, 0, Root, 1, true) };
+  }
+
+  void run(Context &Ctx,
+           const TemporaryFile &SourceBinary,
+           LLVMContainer &TargetsList);
+};
```

Comment:
Currently, contracts are built using indices of arguments of the `run` function.
This bothers me quite a bit. It's easy to get wrong. You add an argument, and do not update the indices.

What if a `Pipe` becomes something like:
```c++
class LiftPipe {
private:
  Context &Ctx;
  const TemporaryFile &SourceBinary;
  LLVMContainer &TargetsList;

public:
  LiftPipe(Context &Ctx, const TemporaryFile &SourceBinary, LLVMContainer &TargetsList) :
    Ctx(Ctx), SourceBinary(SourceBinary), TargetsList(TargetsList) {}

public:
  static constexpr auto Name = "Lift Pipe";
  std::array<AtomicContract, 1> getContract() const {
    return { AtomicContract(Binary, Exactness::Exact, SourceBinary, Root, TargetsList, true) };
  }

  void run();
};
```

Other ideas to get rid of indices are welcome.

Reply:
If you want `getContract` to be static you can use `offsetof(LiftPipe, SourceBinary)` and the like.

Reply:
we can't, the containers are running containers and are materialized while being executed, thus a reference or a pointer provided at build time are not enough.

We can't use types either in the contract because there may be multiple containers with the same type. We can't use names because names depends on the yaml file. 
We don't really have much more to use than indexes.

Reply:
I think you missed my last message about `offsetof`

Reply:
Ah no, wait, I got what you mean.

I think we can shrink the lifetime of the `LiftPipe` object to what previously was the `run` invocation.
Instead of:
```c++
MyLiftPipe.run(A, B);
```
We would have:
```c++
LiftPipe(A, B).run();
```

Then you can make the `getContract` lifetime infinite by making it `static` and using `offsetof` (`LiftPipe::getContract()`).

Reply:
we can't do that neither unless we preclude arbitrary construction of pipes. some pipes may need to be configured

Reply:
What if we drop template parameter pack and push them into a `struct`?

```c++
class LiftPipe {
public:
  static constexpr auto Name = "Lift Pipe";

private:
  int Config;

public:
  struct Arguments {
    const TemporaryFile &SourceBinary;
    LLVMContainer &TargetsList;
  };

  static std::array<AtomicContract, 1> getContract() const {
    return { AtomicContract(Binary, Exactness::Exact, offsetof(Arguments, SourceBinary), Root, offsetof(Arguments TargetsList), true) };
  }

public:
  LiftPipe(int Config) : Config(Config) {}

  void run(Context &Ctx, Arguments &Args);
};
```

Reply:
this version looks extreamly painful w.r.t the original one where the only issue is that one may mess up indexes

Reply:
It is a bit more painful, I agree.

Indices are not obvious though, do you count `Ctx` or not? It'd be nice to make `Context &` mandatory, but then `0` is the *second* argument.
To me, the whole indices thing is a code smell that we're not giving the proper dignity to the container arguments.
We're pretending a pipe is a simple function taking container arguments (which was probably the original goal), but it turns out it's not that simple and we need to come up with workarounds.

`run` is magical in a nice way. But then you *have to* deal with contracts and things become magical in a bad way.

Alternative idea: AFAIU the argument index is just a identifier for a container *within this pipe*. What if we use strings (or maybe `enum`s)?

```c++
class LiftPipe {
public:
  static constexpr auto Name = "Lift Pipe";
  std::array<AtomicContract, 1> getContract() const {
    return { AtomicContract(Binary, Exactness::Exact, "SourceBinary", Root, "TargetsList", true) };
  }

  void run(Context &Ctx,
           const TemporaryFile &SourceBinary,
           LLVMContainer &TargetsList);
};
```

Reply:
we could use strings if we had introspection over arguments names, now we can't since we should use the argument type rather than the argument name and there can be multiple copies of a container type.

if we want secuity we can do something like

```
ArgumentIndex<const TemporaryFile&,1>
```

and have this break if the nth argument does not have the same type as the one provided there